test/integration/fscust_test.go

```diff
@@ -0,0 +1,70 @@
+package integration_test
+
+import (
+	"fmt"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+var fsCustConfigFmt = `{
+  "name": "fs-customizations from local file",
+  "blueprint": {
+    "customizations": {
+      "files": [
+	{
+	  "path": "/etc/file-from-host",
+	  "uri": "file://%s"
+	}
+      ]
+    }
+  }
+}
+`
+
+func TestFileCustomizationFromURI(t *testing.T) {
+	if os.Getuid() != 0 {
+		t.Skip("this test needs to run as root")
+	}
+
+	tmpdir := t.TempDir()
+	canaryString := "testdata"
+	canaryPath := filepath.Join(tmpdir, "canary.txt")
+	err := os.WriteFile(canaryPath, []byte(canaryString), 0600)
+	assert.NoError(t, err)
+	buildConfigPath := filepath.Join(tmpdir, "buildConfig.json")
+	err = os.WriteFile(buildConfigPath, []byte(fmt.Sprintf(fsCustConfigFmt, canaryPath)), 0600)
+	assert.NoError(t, err)
+
+	cmd := exec.Command(
+		"go", "run",
+		// keep tags in sync with BUILDTAGS_CROSS in https://github.com/containers/podman/blob/2981262215f563461d449b9841741339f4d9a894/Makefile#L85
+		"-tags", "containers_image_openpgp exclude_graphdriver_btrfs exclude_graphdriver_devicemapper exclude_graphdriver_overlay",
+		// run build
+		"./cmd/build",
+		"-distro", "centos-10",
+		"-type", "tar",
+		"-config", buildConfigPath,
+		"--output", tmpdir,
+	)
+	cmd.Stdout = os.Stdout
+	cmd.Stderr = os.Stderr
+	cmd.Dir = "../.."
+	err = cmd.Run()
+	assert.NoError(t, err)
+
+	// using ibcli would make this easier as we have --output-name here
+	l, err := filepath.Glob(tmpdir + "/*/archive/*.tar.xz")
+	assert.NoError(t, err)
+	require.Equal(t, 1, len(l))
+
+	// ensure we get the expected content from the locally added file
+	// #nosec G204
+	output, err := exec.Command("tar", "xOf", l[0], "./etc/file-from-host").CombinedOutput()
+	assert.NoError(t, err, "tar output: %s", output)
+	assert.Equal(t, canaryString, string(output))
+}
```

Comment:
This is a bit strange to me, compiling and running the build command from inside a go test.  Let's keep it for now, but I'd prefer to have something a bit cleaner.  It's probably not a lot of work to generate a manifest in code and inspect it for the correct stages, like we do in a lot of cases.

I'm also not sure why we can't have a customization like this in one of the build tests.  We could use a known file (like `os-release`) or set up a mechanism where we create a file on the host before building the test image.
A discussion for another time though.

Reply:
Thanks, I guess we need to sit down and converge on the wider discussion around testing in images long-term. We probably have slightly different views on what "clean" means and how to approach the tests and maybe I should have not written this before we had this discussion.

I would love to hear your thoughts how this would look concretely with the existing framework (I want an end-to-end test, having just a manifest inspection is good but I think its important to test this end-to-end), I'm happy to rework this. In parallel I will try to help kickstart this wider discussion.

Reply:
My 2 cents: This end-to-end testing should be done as part of the boot test of test manifests (see https://github.com/osbuild/images/blob/main/test/scripts/base-host-check.sh). We do this now for other types of customizations. This adds an entirely new approach, which is not ideal from my PoV because it creates friction. I'd suggest using our current approach until we have a consensus on the future form of integration testing of images (vanilla and with customizations).